arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/stm/TQueue.kt

```diff
@@ -0,0 +1,148 @@
+package arrow.fx.coroutines.stm
+
+import arrow.core.Predicate
+import arrow.fx.coroutines.STM
+
+suspend fun <A> STM.newTQueue(): TQueue<A> = TQueue(newTVar(PList.Nil), newTVar(PList.Nil))
+
+/**
+ * A [TQueue] is a transactional unbounded queue which can be written to and read from concurrently.
+ *
+ * ## Creating a [TQueue]
+ *
+ * Creating an empty queue can be done by using either [STM.newTQueue] or [TQueue.new] depending on whether or not you are in
+ *  a transaction or not.
+ *
+ * ## Concurrent reading and writing
+ *
+ * [TQueue] is implemented using two [TVar]'s. One for reading and one for writing.
+ * This effectively means that writing and reading accesses two different variables and thus it never blocks each other.
+ *
+ * > In practice reads have to access the writes variable if they run out of elements to read but this is infrequent.
+ */
+data class TQueue<A> internal constructor(
```

Comment:
Can we not wrap `IQueue` here? And or can/should we improve `IQueue` to incorporate some of these benefits? https://github.com/arrow-kt/arrow-fx/blob/master/arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/IQueue.kt

Reply:
Not as easy no. Because for TQueue to be worth anything we need both lists to be inside `TVar`'s. And yes `IQueue` should also use the linked list implementation that we have here. It is simply better than immutable by copy arraylists. (The gc pressure produced under heavy use is insane...). Another consideration would be a finger tree for `logn` append (amortized to constant time) but that would need some benchmarks first to prove its worth.

Reply:
Oh I see! Now I'm curious to see the benchmark between using a `PList` and `copy arraylists` :D
The `IQueue` is inspired by Scala's Std `Queue`, with a small improvement over it's `size`, but there is no reason why we wouldn't want to improve it futher if we see some good opportuninties.

Reply:
I should set that up. My guess is that a normal linked list will outperform arraylists (maybe even the mutable variant) on every relevant method for queues (`prepend`, `head`, `tail`) and be on par with `reverse`. I am pretty sure that linked lists are optimal of only these 4 methods are concerned and the only thing that can beat those are mutable linked lists, but those come with its own problems. And STM and mutability simply does not work (and shouldn't).

Finger trees could work as well because they have amortized constant time `append` and `tail` (so no need for `reverse` anymore) and constant time `head` (with worst case being `O(logn)`). But the structure there is more complex and thus may be slower. Then again finger trees are amazing for concatenation so if supporting a fast `appendAll` method is desired they might be better again^^

The only thing that the STM implementation cannot do efficiently is track the size. Outside of STM this could be done by tracking appends/reads but in STM this would mean having a shared variable that both reads and writes need to go through and thus it increases contention which is poison for STM's performance. This is also the reason I don't have a size method for `TMap`/`TSet`.

Reply:
Makes perfect sense! Thanks for clarifying further :)